Add homekit_controller tests

Jc2k opened the pull request:
## Description:

This PR pulls the tests out of #20345 and refactors them as requested by @MartinHjelmare. It doesn't add any new features by itself, but it did flush out a bug with the climate implementation.

The initial set of tests covers switch, climate and light platforms. More will be added in subsequent PR's - this is just to see if they pass code review muster in their current form.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):**
home-assistant/home-assistant.io#8307

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Jc2k committed:
Add a test for a homekit_controller switch

Jc2k committed:
Add a test for a homekit_controller lightbulb

Jc2k committed:
Add a test for homekit_controller thermostat

MartinHjelmare commented on homeassistant/components/homekit_controller/climate.py:
```diff
@@ -27,6 +27,8 @@
 # Map of hass operation modes to homekit modes
 MODE_HASS_TO_HOMEKIT = {v: k for k, v in MODE_HOMEKIT_TO_HASS.items()}
 
+DEFAULT_VALID_MODES = list(MODE_HASS_TO_HOMEKIT.keys())
```
```py
DEFAULT_VALID_MODES = list(MODE_HASS_TO_HOMEKIT)
```

MartinHjelmare commented on homeassistant/components/homekit_controller/climate.py:
```diff
@@ -62,8 +64,11 @@ def update_characteristics(self, characteristics):
                     characteristic['value'])
+
+                valid_values = characteristic.get(
+                    'valid-values', DEFAULT_VALID_MODES)
```
This doesn't look right. We're mixing homekit modes with hass modes. DEFAULT_VALID_MODES is a list of hass modes.

MartinHjelmare commented on tests/components/homekit_controller/common.py:
```diff
@@ -0,0 +1,140 @@
+
+    async def get_state(self):
+        """Trigger an entity update() and return the current entity state."""
+        await self.hass.helpers.entity_component.async_update_entity(
```
Why do we need to trigger an update?

MartinHjelmare replied:
We should interact directly with the entity component. If it's a polling entity we should mock time passed. There's a test helper function for that, `async_fire_time_changed`. See manual alarm_control_panel tests for examples.

Jc2k replied:
It's a polling entity (for now). Was calling an update so when i've injected a new value into the fake I need to force a poll to see the outcome of that in the state. So i'll switch to async_fire_time_changed. Could not find how to trigger a poll at all, thx for the pointer!

MartinHjelmare commented on tests/components/homekit_controller/common.py:
```diff
@@ -0,0 +1,140 @@
+        return state
+
+
+async def add_fake_homekit_accessory(hass, services):
```
It's not clear from the name that this will also set up the component. Maybe break that out, or rename this function?

MartinHjelmare commented on tests/components/homekit_controller/test_climate.py:
```diff
@@ -0,0 +1,77 @@
+    helper.characteristics[HEATING_COOLING_CURRENT].value = 1
+    helper.characteristics[HEATING_COOLING_TARGET].value = 1
+
+    climate_1 = await helper.get_state()
```
We usually just use the variable `state` here. That's more clear so we don't think it's an entity.

Jc2k committed:
Changes from review

MartinHjelmare commented on tests/components/homekit_controller/common.py:
```diff
@@ -88,16 +90,19 @@ def __init__(self, hass, entity_id, pairing, accessory):
-            self.entity_id)
+    async def poll_and_get_state(self):
+        """Trigger a time based poll and return the current entity state."""
+        next_update = dt_util.utcnow() + timedelta(seconds=60)
```
Make sure to patch `dt_util.utcnow` in the test too to take complete control of time.

Jc2k replied:
When I do that it doesn’t trigger a poll for some reason. It only works when I don’t patch it.

MartinHjelmare replied:
Make sure to patch in the correct place, and also watch for any throttle. Don't remember if that is in play for this component.

MartinHjelmare replied:
Make sure to patch time already before setting up the component. It's when the entity is added that the clock starts.

Jc2k committed:
Patch utcnow to known time in HK tests

Jc2k committed:
Neater fixture use per review

Jc2k commented on tests/components/homekit_controller/conftest.py:
```diff
@@ -0,0 +1,15 @@
+    """Freeze time at a known point."""
+    start_dt = datetime.datetime(2019, 1, 1, 0, 0, 0)
+    patcher = mock.patch('homeassistant.util.dt.utcnow', return_value=start_dt)
+    patcher.start()
```
I forgot you could yield from a pytest.Fixture! Good call, done.

MartinHjelmare commented on script/gen_requirements_all.py:
```diff
@@ -32,7 +32,6 @@
     'i2csense',
     'credstash',
     'bme680',
-    'homekit',
```
Have the extra dependencies for homekit been removed?
See note here:
https://www.home-assistant.io/components/homekit_controller/

I think that was why we had commented it here.

Jc2k replied:
Indeed thats out of date, those are no longer required! They were worked around so I could turn on tests here and to make it easier to install on mac / raspbian.

Jc2k replied:
(I'll figure out the docs repository tomorrow and try and submit a PR)

Jc2k replied:
PR: https://github.com/home-assistant/home-assistant.io/pull/8307

MartinHjelmare reviewed (commented):
What's the test coverage now? Should we remove the tested modules from coveragerc?

Jc2k commented:
I'd certainly like to. It's definitely not 100%. However it's certainly better than my last attempt at tests, so thank you for pushing me to face the more challenging approach! Let's see...

```
$ coverage run -m pytest tests/components/homekit_controller/
$ coverage report | grep homekit_controller
homeassistant/components/homekit_controller/__init__.py                  181     57    69%
homeassistant/components/homekit_controller/alarm_control_panel.py        52     52     0%
homeassistant/components/homekit_controller/climate.py                    64      0   100%
homeassistant/components/homekit_controller/cover.py                     154    154     0%
homeassistant/components/homekit_controller/light.py                      74      9    88%
homeassistant/components/homekit_controller/lock.py                       50      5    90%
homeassistant/components/homekit_controller/switch.py                     38      0   100%
```

I'm in the middle of writing a test for lock which i planned to send seperately (it adds an extra test helper), i imagine cover and alarm control panel will get to a similar level at the same time. In __init__.py the biggest chunk of untested code is probably the pairing code. 

I was going to see where things were at after i'd merged this and added the other set of tests (covers, locks, alarm control panels). 

So options are:

 * Remove homekit_controller from .coveragerc entirely. Coverage will probably decrease, and possibly enough to give the PR a red flag(?)
 * Remove switch + climate only. This will replace the glob with an explicit list and mean that new entities have to have tests.
 * Same but switch, climate and light.
 * Leave it till i've added more tests.

MartinHjelmare commented:
Let's leave it until more tests are added. Each module not in `.coveragerc` needs to be over 95% coverage.

MartinHjelmare reviewed (approved):
Great!
